src/transformers/cache_utils.py

```diff
@@ -320,3 +322,91 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+    
+    # TODO extra the batchsize in the generate method.
+    def __init__(self, config: PretrainedConfig, num_layers, max_batch_size, max_sequence_length, num_heads, hidden_dim, dtype=torch.float16) -> None:
+        super().__init__()
+        self.num_layers = num_layers
+        self.max_batch_size = max_batch_size
+        self.max_sequence_length = max_sequence_length
+        self.head_dim = hidden_dim // num_heads
+        self.num_heads = num_heads
+        self.shape = (max_batch_size, max_sequence_length, hidden_dim // num_heads,  num_heads)
+        self.dtype = dtype # Property?
+        
+        # TODO device meta? 
+        self.key_cache: List[torch.Tensor] = [torch.zeros(max_batch_size, num_heads, max_sequence_length, self.head_dim, dtype=dtype) for _ in range(num_layers)]
+        self.value_cache: List[torch.Tensor] = [torch.zeros(max_batch_size, num_heads, max_sequence_length, self.head_dim, dtype=dtype) for _ in range(num_layers)]
+        
+        # FIXME our format should be the followingm, but most of our nlp model apply transpose operation on the k and v so we can't 
+        # self.key_cache: List[torch.Tensor] = [torch.zeros(max_batch_size, max_sequence_length, num_heads, self.head_dim, dtype=dtype) for _ in range(num_layers)]
+        # self.value_cache: List[torch.Tensor] = [torch.zeros(max_batch_size, max_sequence_length, num_heads, self.head_dim, dtype=dtype) for _ in range(num_layers)]
+        self.seen_tokens = 0  # Used in `generate` to keep tally of how many tokens the cache has seen
+
+
+    def update(
+        self,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cache_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Updates the cache with the new `key_states` and `value_states` for the layer `layer_idx`.
+
+        Parameters:
+            key_states (`torch.Tensor`):
+                The new key states to cache.
+            value_states (`torch.Tensor`):
+                The new value states to cache.
+            layer_idx (`int`):
+                The index of the layer to cache the states for.
+            cache_kwargs (`Dict[str, Any]`, `optional`):
+                Additional arguments for the cache subclass. No additional arguments are used in `DynamicCache`.
+
+        Return:
+            A tuple containing the updated key and value states.
+        """
+        if self.seen_tokens == 0:
+            self.key_cache[layer_idx] = self.key_cache[layer_idx].to(key_states.device)
+            self.value_cache[layer_idx] = self.value_cache[layer_idx].to(value_states.device)
+
+        # Update the cache
+        if len(self.key_cache) + 1 == self.max_sequence_length:
+            # let's overwrite and roll the cache to support going beyond?
+            self.key_cache[layer_idx][0] = key_states
+            self.key_cache[layer_idx] = torch.roll(self.key_cache[layer_idx],-1,0)
```

Comment:
Hmm I'd not allow this. `torch.roll` doesn't work well with torch.compile as it dynamically allocates new memory. I'd throw an error instead if the key_cache length is larger then max_length

Reply:
Agreed!